core/src/main/java/com/mongodb/hadoop/output/MongoOutputCommitter.java

```diff
@@ -102,12 +105,12 @@ public void commitTask(final TaskAttemptContext taskContext) {
                 int mwType = inputStream.readInt();
                 if (MongoWritableTypes.BSON_WRITABLE == mwType) {
                     bw.readFields(inputStream);
-                    bulkOp.insert(new BasicDBObject(bw.getDoc().toMap()));
+                    insert.putAll(bw.getDoc().toMap());
```

Comment:
`putAll` won't remove the fields that were in `insert` from a previous document, so if not all documents have exactly the same schema, then later documents will have extraneous fields in them. A similar problem applies to `query` and `modifiers` below as well. If you made these all `BasicDBObject`s instead of `DBObject`s, you could call the `clear` method before `putAll`.